
---
title: "R_Assignment"
author: "Cecilia Hodson"
date: '2021-03-14'
output:
  html_document: default
  pdf_document: default
---

# Part I: Data Inspection

### Install packages n' stuff
```{r}
if (!require("tidyverse")) install.packages("tidyverse")
library(tidyverse)
if (!require("ggplot2")) install.packages("ggplot2")
library(ggplot2)
if (!require("gtools")) install.packages("gtools")
library(gtools)
if (!require("reshape2")) install.packages("reshape2")
library(reshape2)
if (!require("httr")) install.packages("httr")
library(httr)
```
### Looking at the files:
```{r}
file.size("fang_et_al_genotypes.txt")
file.size("snp_position.txt")
```
### Creating directories for future files:
```{r}
dir.create("maize_files")
dir.create("teosinte_files")
dir.create("graphs")
```

### Creating data frames for .txt files:
```{r}
# I guess you need to import the data? I had a buddy run this for me and it wouldn't work for him until he imported these two .txt files. 
fang_genotypes <- read.delim("fang_et_al_genotypes.txt")
snp_positions <- read.delim("snp_position.txt")
```
### Looking at the dataframes:
```{r}
nrow(fang_genotypes)
ncol(fang_genotypes)

nrow(snp_positions)
ncol(snp_positions)

```
## Data Processing

### Pulling genotype groups:
```{r}
maize <- fang_genotypes[fang_genotypes$Group =="ZMMIL" | fang_genotypes$Group == "ZMMMR" | fang_genotypes$Group == "ZMMLR",]
teosinte <- fang_genotypes[fang_genotypes$Group == "ZMPIL" | fang_genotypes$Group == "ZMPJA" | fang_genotypes$Group =="ZMPBA",]
```
### Transposing genotypes:
```{r}
transposed_maize <- as.data.frame(t(maize[,-1]))
colnames(transposed_maize) <- maize$Sample_ID
transposed_teosinte <- as.data.frame(t(teosinte[,-1]))
colnames(transposed_teosinte) <- teosinte$Sample_ID
```
### Pulling SNP Colums:
```{r}
snp_cols <- snp_positions[, c(1, 3, 4)]
```
### Merging Files:
```{r}
# initial merging
merged_maize <- merge(x = snp_cols, y = transposed_maize, by.x = "SNP_ID", by.y = "row.names", all.y = TRUE)
merged_teosinte <-  merge(x = snp_cols, y= transposed_teosinte, by.x = "SNP_ID", by.y = "row.names", all.y = TRUE)

# remove certain rows that are not data (i.e. "Group" and whatever "JG-OTU" is...)
merged_maize <- merged_maize[-c(14, 16), ]
merged_teosinte <- merged_teosinte[-c(14, 16), ]

# removing undesirable values(i.e., "multiple" & "unknown")
merged_maize <- merged_maize[!(merged_maize$Chromosome == "multiple"), ]
merged_maize <- merged_maize[!(merged_maize$Chromosome == "unknown"), ]
merged_maize <- merged_maize[!(merged_maize$Position == "unknown"), ]

merged_teosinte <- merged_teosinte[!(merged_teosinte$Chromosome == "multiple"), ]
merged_teosinte <- merged_teosinte[!(merged_teosinte$Chromosome == "unknown"), ]
merged_teosinte <- merged_teosinte[!(merged_teosinte$Position == "unknown"), ]

# sorting position column
merged_maize$Position <- as.numeric(merged_maize$Position)
merged_teosinte$Position <- as.numeric(merged_teosinte$Position)

merged_maize <- merged_maize[order(merged_maize$Position), ]
merged_teosinte <- merged_teosinte[order(merged_teosinte$Position), ]

```
### Sorting by Chromosome Number:
```{r}
# for maize
maize_split_list <- split(merged_maize, merged_maize$Chromosome, drop = FALSE)

maize_chr_1 <- maize_split_list$"1" 
maize_chr_2 <- maize_split_list$"2"
maize_chr_3 <- maize_split_list$"3"
maize_chr_4 <- maize_split_list$"4"
maize_chr_5 <- maize_split_list$"5"
maize_chr_6 <- maize_split_list$"6"
maize_chr_7 <- maize_split_list$"7"
maize_chr_8 <- maize_split_list$"8"
maize_chr_9 <- maize_split_list$"9"
maize_chr_10 <- maize_split_list$"10"


# for teosinte
teo_split_list <- split(merged_teosinte, merged_teosinte$Chromosome, drop = FALSE)

teo_chr_1 <- teo_split_list$"1"
teo_chr_2 <- teo_split_list$"2"
teo_chr_3 <- teo_split_list$"3"
teo_chr_4 <- teo_split_list$"4"
teo_chr_5 <- teo_split_list$"5"
teo_chr_6 <- teo_split_list$"6"
teo_chr_7 <- teo_split_list$"7"
teo_chr_8 <- teo_split_list$"8"
teo_chr_9 <- teo_split_list$"9"
teo_chr_10 <- teo_split_list$"10"

# Sorting by decreasing position (reversed):

maize_reverse <- merged_maize[order(merged_maize$Position, decreasing=T),]
teosinte_reverse <- merged_teosinte[order(merged_teosinte$Position, decreasing=T),]


maize_reverse[] <- lapply(maize_reverse, as.character)
maize_reverse[maize_reverse == "?/?"] <- "-/-"
teosinte_reverse[] <- lapply(teosinte_reverse, as.character)
teosinte_reverse[teosinte_reverse == "?/?"] <- "-/-"


maize_split_reverse <- split(maize_reverse, maize_reverse$Chromosome)

maize_reverse_chr1 <- maize_split_reverse$"1"
maize_reverse_chr2 <- maize_split_reverse$"2"
maize_reverse_chr3 <- maize_split_reverse$"3"
maize_reverse_chr4 <- maize_split_reverse$"4"
maize_reverse_chr5 <- maize_split_reverse$"5"
maize_reverse_chr6 <- maize_split_reverse$"6"
maize_reverse_chr7 <- maize_split_reverse$"7"
maize_reverse_chr8 <- maize_split_reverse$"8"
maize_reverse_chr9 <- maize_split_reverse$"9"
maize_reverse_chr10 <- maize_split_reverse$"10"


teosinte_split_rev <- split(teosinte_reverse, teosinte_reverse$Chromosome)

teo_reverse_chr1 <- teosinte_split_rev$"1"
teo_reverse_chr2 <- teosinte_split_rev$"2"
teo_reverse_chr3 <- teosinte_split_rev$"3"
teo_reverse_chr4 <- teosinte_split_rev$"4"
teo_reverse_chr5 <- teosinte_split_rev$"5"
teo_reverse_chr6 <- teosinte_split_rev$"6"
teo_reverse_chr7 <- teosinte_split_rev$"7"
teo_reverse_chr8 <- teosinte_split_rev$"8"
teo_reverse_chr9 <- teosinte_split_rev$"9"
teo_reverse_chr10 <- teosinte_split_rev$"10"
```





# Part II: Visualizing Data

### Install required packages:
```{r}
if (!require("ggplot2")) install.packages("ggplot2")
library(ggplot2)
if (!require("reshape2")) install.packages("reshape2")
library(reshape2)
```


## SNPs Per Chromosome:
  - This is for all groups, not just the maize and teosinte.
```{r}
# First, create a merged genotypes file:
genotypes <- as.data.frame(t(fang_genotypes[, -1]))
merged_genotypes <- merge(x = snp_cols, y = genotypes, by.x = "SNP_ID", by.y = "row.names", all.y = TRUE)
merged_genotypes <- merged_genotypes[-c(14, 16), ]
merged_genotypes$Chromosome <- factor(merged_genotypes$Chromosome, levels = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown", "NA"))


# Making the plot and adding a little pink in the plot, cause I like pink :)
ggplot(merged_genotypes, aes(x = Chromosome, fill = Chromosome)) + geom_bar(stat=) + ggtitle("SNPs per Chromosome") + theme(
  axis.text.x = element_text(colour = "black", angle = 45),
  axis.title.x = element_text(colour = "black", size=rel(2)),
  axis.title.y = element_text(colour = "black", angle=90, size=rel(2)),
  panel.background = element_rect(fill="white"),
  plot.background = element_rect(fill="pink"))  

ggsave(filename = "graphs/SNPs per Chromosome.png", device = "png")
```

### Distibution of SNPs per Chromosome
```{r}
# Forcing the data to be what we want it to be
merged_genotypes <-  merged_genotypes[!(merged_genotypes$Position == "unknown"), ]
merged_genotypes <-  merged_genotypes[!(merged_genotypes$Position == "multiple"), ]
merged_genotypes$Position <- as.numeric(merged_genotypes$Position)


# Making the plot itself
ggplot(merged_genotypes) + geom_density(mapping = aes(x=Position, colour=Chromosome), alpha=0.5) + ggtitle("Density of Position by Chromosome") + theme(
    axis.text = element_text(colour = "purple"),
    axis.title.x = element_text(colour = "purple"),
    axis.title.y = element_text(colour = "purple", angle=90, size=rel(2))
    )

ggsave(filename = "graphs/Density of Position by Chromosome.png", device = "png")
```


## Missing Data and Heterozygosity:
```{r}
# Grabbing the SNPS
zygosity_genotypes_extended <- fang_genotypes[fang_genotypes$Group == "ZMMIL" | fang_genotypes$Group == "ZMMLR" | fang_genotypes$Group == "ZMMMR" | fang_genotypes$Group == "ZMPBA" | fang_genotypes$Group == "ZMPIL" | fang_genotypes$Group == "ZMPJA",]

zygosity_genotypes <- melt(zygosity_genotypes_extended, measure.vars = colnames(fang_genotypes)[4:986])
colnames(zygosity_genotypes)[4:5] <- c("SNP_ID", "Homozygous")

# Applying that "TRUE/FALSE" property to the newly created columns:
zygosity_genotypes[zygosity_genotypes == "A/A"] <- TRUE
zygosity_genotypes[zygosity_genotypes == "C/C"] <- TRUE
zygosity_genotypes[zygosity_genotypes == "G/G"] <- TRUE
zygosity_genotypes[zygosity_genotypes == "T/T"] <- TRUE

zygosity_genotypes[zygosity_genotypes == "A/C"] <- FALSE
zygosity_genotypes[zygosity_genotypes == "A/G"] <- FALSE
zygosity_genotypes[zygosity_genotypes == "A/T"] <- FALSE
zygosity_genotypes[zygosity_genotypes == "C/G"] <- FALSE
zygosity_genotypes[zygosity_genotypes == "C/T"] <- FALSE
zygosity_genotypes[zygosity_genotypes == "G/T"] <- FALSE

zygosity_genotypes[zygosity_genotypes == "?/?"] <- NA

# Time to sort!
zygosity_genotypes <- arrange(zygosity_genotypes, Sample_ID, Group)

# Time to plot!
ggplot(data = zygosity_genotypes) +
    geom_bar(mapping = aes(x = Sample_ID, fill = Homozygous), stat = "count") +
    ggtitle(label = "SNP Zygosity") +
    ylab(label = "Number of SNPs") +
    theme(
        plot.title = element_text(size=rel(2)),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

ggsave(filename = "graphs/Heterozygosity and Missing Data.png", device = "png")
```

## Distribution of Groups
```{r}
# Showing a bar graph of number of samples belonging to each group (there seems to be a lot of maize...)
ggplot(fang_genotypes, aes(x = Group)) + geom_bar(stat=, fill = "pink", position="dodge",) + ggtitle("Data Group Distribution") + theme(
        axis.text.x = element_text(colour = "black", angle = 45),
        axis.title.x = element_text(colour = "black", size=rel(2)),
        axis.title.y = element_text(colour = "black", angle=90, size=rel(2)),
        panel.background = element_rect(fill="white"),
        plot.background = element_rect(fill="white"))

ggsave(filename = "graphs/Data Group Distribution.png", device = "png")

# Removing large groups:
fang_no_groups <- fang_genotypes[fang_genotypes$Group =="TRIPS" | fang_genotypes$Group == "ZDIPL" | fang_genotypes$Group == "ZLUXR" | fang_genotypes$Group =="ZMHUE" | fang_genotypes$Group =="ZMMMR" | fang_genotypes$Group =="ZMPIL" | fang_genotypes$Group =="ZMPJA" | fang_genotypes$Group =="ZMXCH" | fang_genotypes$Group =="ZMXIL" | fang_genotypes$Group =="ZMXNO" | fang_genotypes$Group =="ZMXNT" | fang_genotypes$Group =="ZPERR", ]


# Plot without large groups
ggplot(fang_no_groups, aes(x = Group)) + geom_bar(stat=, fill = "pink", position="dodge",) + ggtitle("Data Group Distribution 2") + theme(
    axis.text.x = element_text(colour = "black", angle = 45),
    axis.title.x = element_text(colour = "black", size=rel(2)),
    axis.title.y = element_text(colour = "black", angle=90, size=rel(2)),
    panel.background = element_rect(fill="white"),
    plot.background = element_rect(fill="white"))

ggsave(filename = "graphs/Data Group Distribution 2.png", device = "png")
```










